OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -322,6 +216,174 @@ void TracerThread::Run(
     close(fd);
   }
   fds_to_ring_buffer.clear();
+
+  for (int uret_probe_fd : uret_probes_fds) {
+    perf_event_disable(uret_probe_fd);
+    close(uret_probe_fd);
+  }
+}
+
+void TracerThread::ProcessContextSwitchEvent(const perf_event_header& header,
+                                             PerfEventRingBuffer& ring_buffer) {
+  ContextSwitchPerfEvent event;
+  ring_buffer.ConsumeRecord(header, &event.ring_buffer_record);
+  ++sched_switch_count_;
+  pid_t tid = event.GetTid();
+  uint16_t cpu = static_cast<uint16_t>(event.GetCpu());
+  uint64_t time = event.GetTimestamp();
+
+  if (event.IsSwitchOut()) {
+    listener_->OnContextSwitchOut(ContextSwitchOut(tid, cpu, time));
+  } else {
+    listener_->OnContextSwitchIn(ContextSwitchIn(tid, cpu, time));
+  }
+}
+
+void TracerThread::ProcessContextSwitchCpuWideEvent(
+    const perf_event_header& header, PerfEventRingBuffer& ring_buffer) {
+  SystemWideContextSwitchPerfEvent event;
+  ring_buffer.ConsumeRecord(header, &event.ring_buffer_record);
+
+  if (event.GetPrevTid() != 0) {
+    ContextSwitchOut context_switch_out{event.GetPrevTid(),
+                                        static_cast<uint16_t>(event.GetCpu()),
+                                        event.GetTimestamp()};
+    listener_->OnContextSwitchOut(context_switch_out);
```

Comment:
I know that this class doesn't make a lot of sense with listener_ being nullptr, but are we removing all the checks for nullptr?

Reply:
Could we just have a single check / assert somewhere?